api/src/main/java/ai/djl/nn/transformer/TransformerBaseBlock.java

```diff
@@ -0,0 +1,256 @@
+     *     in one line
+     */
+    protected <P extends Parameter> P addParameter(
+            P parameter, Function<Shape[], Shape> shapeCallback) {
+        parameters.put(parameter.getName(), parameter);
+        parameterShapeCallbacks.put(parameter.getName(), shapeCallback);
+        return parameter;
+    }
+
+    @Override
+    public Shape getParameterShape(String name, Shape[] inputShapes) {
+        Function<Shape[], Shape> callback = parameterShapeCallbacks.get(name);
+        if (callback == null) {
+            Parameter parameter = parameters.get(name);
+            if (parameter == null) {
+                throw new IllegalArgumentException(
+                        "No parameter named " + name + " found in this block.");
+            } else {
+                throw new IllegalStateException(
+                        "No shape initializer for parameter "
+                                + name
+                                + "found. "
+                                + "Either pass an initializer for the shape when adding the parameter or override "
+                                + "getParameterShape in the subclass.");
+            }
+        }
+        return callback.apply(inputShapes);
+    }
+
+    @Override
+    public BlockList getChildren() {
+        return new BlockList(children);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public Shape[] initialize(NDManager manager, DataType dataType, Shape... inputShapes) {
+        beforeInitialize(inputShapes);
+        for (Parameter parameter : getDirectParameters()) {
+            parameter.initialize(manager, dataType, inputShapes);
+        }
+        initializeChildBlocks(manager, dataType, inputShapes);
+        return getOutputShapes(manager, inputShapes);
+    }
+
+    /**
+     * Initializes the Child blocks of this block.
+     *
+     * @param manager the manager to use for initialization
+     * @param dataType the requested data type
+     * @param inputShapes the expected input shapes
+     */
+    public abstract void initializeChildBlocks(
+            NDManager manager, DataType dataType, Shape... inputShapes);
+
+    /**
+     * Default implementation of predict that calls forward.
+     *
+     * @param parameterStore the parameter store
+     * @param inputs the input NDList
+     * @param params optional parameters
+     * @return the prediction result, same as from forward pass
+     */
+    @Override
+    public NDList predict(
+            ParameterStore parameterStore, NDList inputs, PairList<String, Object> params) {
+        return forward(parameterStore, inputs, params);
+    }
+
+    @Override
+    public List<Parameter> getDirectParameters() {
+        return new ArrayList<>(parameters.values());
+    }
+
+    @Override
+    public void saveParameters(DataOutputStream os) throws IOException {
+        os.write(version);
+        for (Parameter parameter : parameters.values()) {
+            parameter.save(os);
+        }
```

Comment:
You may need to save/load some instance variables as well depending on the block (e.g. Embedding). Maybe add an additional saveInstanceVariables/load methods that can be overridden but default to nothing?

Reply:
Yes, I thought that as well, as looking at other blocks I saw that they store additional state. If nobody else is working on the merge and you are OK with it, I'd volunteer to do it and add methods for saving / loading instance state. Also I'd retrofit the existing blocks.

Reply:
That sounds great. Go ahead and update it